Fixes & Additions to Progressbar.

Speiger opened the pull request:
-Optional: Added Default Interface Methods for Creating Progress Style. (can be removed if wanted but shouldn't)
-Suggestion: Styles should be part of the API so defaults or caches can be created
-Fixed: Progressbar was able to overflow (< vs <= in for loop when full).
-Fixed: Progressbars are now Localizable because server can not use language
-Added: Progressbar Text can now be aligned (Yes it has usecases)
-Added: Helper functions for "color" and "bounds" that reduce code but do the same.
-Added: Ability to clone styles

My first suggestion. The wall of text for "Creation of styles" can be removed if really unwanted.
But Styles should be createable even during game load so you can create templates on what you need and add cloning.
Especially if there is a lot of  small replication since instances are shared it reduces memory usage garbage creation and allows to create small changes.

It makes sense with certain design patterns at least. Tell me your thoughts. I can do edits based on what is accepted and not.
Once this is through next patch will be created.

Edit: Tested the patch works with previous versions and just adds a lot of things.
Edit2: Yes the IProgressStyle imports its implementation. But styles should in general be in the API even their impl

Speiger committed:
Fixes & Additions to Progressbar.

-Optional: Added Default Interface Methods for Creating Progress Style. (can be removed if wanted but shouldn't)
-Suggestion: Styles should be part of the API so defaults or caches can be created
-Fixed: Progressbar was able to overflow (< vs <= in for loop when full).
-Fixed: Progressbars are now Localizable because server can not use language
-Added: Progressbar Text can now be aligned (Yes it has usecases)
-Added: Helper functions for "color" and "bounds" that reduce code but do the same.

Speiger commented:
@McJty why was it denied?

McJty commented:
Oh that was accidental. Apparently the PR that I wanted to close was already closed.
Sorry

McJty commented:
I'll look over this soon

Speiger commented:
Ok i was really confused there.

McJty commented:
A few remarks:

- Please remove your usage of OnlyIn. That tag should never be used by mods. It's intended for Forge internal usage. It's also never needed
- Don't use AWT Color. Best to use our own Color. Using AWT can potentially cause issues as it's specifically forbidden in combination with Minecraft/JWGL
- I do NOT want functions like createBorderlessColored and createAll and all related functions. The way I designed the builder style api is to avoid needing those functions with tons of parameters and tons of functions. I really don't want to go to that kind of API
- The other changes look ok but I want to review them better with a smaller PR that fixes the above issues

Thanks

Speiger commented:
Eh yeah answers to that

- Only in is not a forge only internal class It is a side detector that allows to remove functions if they are only a specific side. So if i use in a Block TextureAtlasSprite (for a reason) if i don't add OnlyIn the server crashes. Actually its a wonder that your mod didn't crash because of it. Unless you know something that i don't know.
- Yeah you need to elaborate on the color part. Because this sounds like you don't know java api classes. MC actually uses the AWT Color Standard right out of the gate and OpenGL supports it too since its a Open Programming Language Standard. (RGBA)
On top of that Ides support color with "highlight/previewing" it allowing to visualize it much better
- Well the one thing the interface should provide is "createDefault" for the reason that people otherwise have to use the API impl just to create variables because the API is not exposing the creation of Styles in general. I get that a lot of the static functions could be cut because they are to MUCH. But the API needs at the very least then helper functions to reduce code size on the creation.
- "Smaller?" this is the smallest that wouldn't require 3 branches just to get the same  content through. Like O.o

Speiger committed:
Requested Changes

-Removed: Most of the Creation Methods, kept a couple more then requested. These are the most common use-cases that user will use. Only Essentials.
-Added: color & borderless color functions that allow to not have 500 characters just to set a basic custom colored progressbar. Or go multiline to keep it readable.

Speiger committed:
Requested Changes

-Removed: Most of the Creation Methods, kept a couple more then requested. These are the most common use-cases that user will use. Only Essentials.
-Added: color & borderless color functions that allow to not have 500 characters just to set a basic custom colored progressbar. Or go multiline to keep it readable.

McJty commented:
- Forge can use OnlyIn. That's fine for forge. But note the comment that comes with OnlyIn:
 * <p>This is generally meant for internal Forge and FML use only
 * and modders should avoid its use whenever possible.</p>

- Your PR is already much better but I'm still not a fan of your helper functions. They are not needed. It doesn't save that much and it's much more readable if you be more explicit. But for the colors maybe. Have to think about it

Anyway, I'll probably merge it as such later this evening and make the rest of the changes manually. I'll probably keep the color conveniance methods. I'll think about the rest.

Thanks

Speiger committed:
Added Style getter of a progress Element.

-Usecase: Allows to edit the progressbar after its creation if someone wants to change the base

Speiger commented:
Small edit added a style access because if you want to override how oneprobe handles one of your progressbars you either have to entirely replace the object and if it is style difference that would remove that need. (Casting still required)
